### PR TITLE
Base 64 decoder, reject input when unused bits are not 0

### PR DESCRIPTION
**Pull request by buyaa-n**
When the input is not multiple of 3 some bits of the encoded values are not used, Base64 encoder sets those bits to 0, but decoder currently doesn't check those bits and allows any combination of values. Therefore multiple input could decoded to same value, for example when the input is 1 byte character `'A'`, encoder encodes it to 2 base64 characters and 2 padding `"QQ=="` , the last 4 bits of the 2nd `Q` is not used and set to 0s, but decoder doesn't validate that and allows 2^4 = 16 values decoded into same value, for example: "QQ==", "QR==", "QS==", "QT==", "QV==", "QU==", "QW==", "QX==", "QY==", "QZ==", "Qa==", "Qb==", "Qc==", "Qd==", "Qe==", "Qf" will be decoded to a same value, 65, ascii of 'A'.

The [spec mentions](https://datatracker.ietf.org/doc/html/rfc4648#section-3.5) that unused bits MUST be set to zero by conforming encoders. It also mentions that decoders may reject an input if pad bits have not been set to zero. We don't see any reason to keep allowing non-zero value for those other combinations that produce same result when encoders are expected to produce only one value. 

This doesn't seem to be a breaking change, my quick research did not find any encoder, that produces output that doesn't set unused bits to 0. Though it could break tests that randomly generates Base64 encoded text.

Further we should fix this for `Convert.FromBase64XYZ(...)` overloads

Related to https://github.com/dotnet/runtime/issues/50233#issuecomment-807241266

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs [hunk 1]
```diff
@@ -9,23 +9,29 @@ internal static partial class Base64Helper
...
                         length += base64Text.Length;
+                        lastChar = base64Text[base64Text.Length - 1];
                         break;
                     }
 
                     length += index;
+                    if (index > 0)
+                    {
+                        lastChar = base64Text[index - 1];
+                    }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs:34]: Are there any perf consequences to these changes?

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs:34]: Both APIs did not show noticeable impact with the perf tests I run, but the input doesn't include whitespaces, I will try with an input with whitespaces

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs:34]: Around 0.03 percent degradation with whitespace

| Method                       | Toolchain           | Mean       | Error     | StdDev     | Median     | Ratio | RatioSD |
|----------------------------- |-------------------- |-----------:|----------:|-----------:|-----------:|------:|--------:|
| Base64IsValid                | \lastPR\corerun.exe |   5.734 us | 0.0576 us |  0.0511 us |   5.729 us |  1.01 |    0.01 |
| Base64IsValid                | \main\corerun.exe   |   5.665 us | 0.0159 us |  0.0133 us |   5.665 us |  1.00 |    0.00 |
|                              |                     |            |           |            |            |       |         |
| Base64IsValidWithWhiteSpaces | \lastPR\corerun.exe | 173.792 us | 0.5332 us |  0.4988 us | 173.837 us |  1.03 |    0.00 |
| Base64IsValidWithWhiteSpaces | \main\corerun.exe   | 169.125 us | 0.4375 us |  0.3879 us | 169.179 us |  1.00 |    0.00 |
|                              |                     |            |           |            |            |       |         |
| Base64Decode                 | \lastPR\corerun.exe |   6.036 us | 0.0236 us |  0.0221 us |   6.032 us |  1.00 |    0.01 |
| Base64Decode                 | \main\corerun.exe   |   6.044 us | 0.0276 us |  0.0231 us |   6.031 us |  1.00 |    0.00 |
|                              |                     |            |           |            |            |       |         |
| Base64DecodeInPlace          | \lastPR\corerun.exe | 104.810 us | 2.0912 us |  5.4724 us | 103.150 us |  0.94 |    0.10 |
| Base64DecodeInPlace          | \main\corerun.exe   | 112.553 us | 3.9723 us | 11.3972 us | 109.000 us |  1.00 |    0.00 |

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs [hunk 2]
```diff
@@ -216,7 +216,7 @@ internal static partial class Base64Helper
                 }
                 else
                 {
-                    if (i0 < 0)
+                    if (i0 < 0 || (i0 & 0xF000) != 0) // also check 4 unused lower bits are set to 0
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:219]: i0 < 0 is really just checking whether the upper bit is set. So could this be collapsed into a single condition?
```C#
if ((i0 & 0x8000F000) != 0)
```
?

Same for the other similar one later.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs [hunk 3]
```diff
@@ -169,10 +177,19 @@ public int DecodeValue(char value)
...
             public bool IsEncodingPad(byte value) => value == EncodingPad;
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            public bool ValidateAndDecodeLength(int length, int paddingCount, out int decodedLength)
+            public bool ValidateAndDecodeLength(byte lastChar, int length, int paddingCount, out int decodedLength)
             {
                 if (length % 4 == 0)
                 {
+                    int decoded = default(Base64DecoderByte).DecodingMap[lastChar];
+                    if (paddingCount == 1 && (decoded & 0x03) != 0 ||
+                        paddingCount == 2 && (decoded & 0x0F) != 0)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs:186]: Please add parens here to make the precedence clearer

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlValidator.cs [hunk 4]
```diff
@@ -112,6 +114,15 @@ public bool ValidateAndDecodeLength(int length, int paddingCount, out int decode
 
                 decodedLength = (length >> 2) * 3 + (remainder > 0 ? remainder - 1 : 0) - paddingCount;
 #endif
+                int decoded = default(Base64DecoderByte).DecodingMap[lastChar];
+                if ((remainder == 3 || paddingCount == 1) && (decoded & 0x03) != 0 ||
+                    (remainder == 2 || paddingCount == 2) && (decoded & 0x0F) != 0)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlValidator.cs:119]: Ditto

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs]: Good point, thanks

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs [hunk 5]
```diff
@@ -9,23 +9,29 @@ internal static partial class Base64Helper
...
                     int index = validatable.IndexOfAnyExcept(base64Text);
                     if ((uint)index >= (uint)base64Text.Length)
                     {
                         length += base64Text.Length;
+                        lastChar = base64Text[base64Text.Length - 1];
                         break;
                     }
 
                     length += index;
+                    if (index > 0)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs:31]: Is this really a check for index != 0? We just validated that it's not negative in the previous if check.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs [hunk 6]
```diff
@@ -426,7 +426,7 @@ internal static unsafe OperationStatus DecodeFromUtf8InPlace<TBase64Decoder>(TBa
                 }
                 else
                 {
-                    if (i0 < 0)
+                    if ((i0 & 0x8000F000) != 0) // if negative or 4 unused bits are not 0.
```
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:429]: This one is correct, the above is for last 2 bytes

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs (hunk 5 above)
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64ValidatorHelper.cs:31]: Yes, that is true

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs [hunk 7]
```diff
@@ -415,7 +415,7 @@ internal static unsafe OperationStatus DecodeFromUtf8InPlace<TBase64Decoder>(TBa
 
                     i0 |= i2;
 
-                    if (i0 < 0)
+                    if ((i0 & 0x800000c0) != 0) // if negative or 4 unused bits are not 0.
```
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:418]:
```suggestion
                    if ((i0 & 0x800000c0) != 0) // if negative or 2 unused bits are not 0.
```

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs [hunk 8]
```diff
@@ -200,7 +200,7 @@ internal static partial class Base64Helper
...
```
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:203]:
```suggestion
                    if ((i0 & 0x800000c0) != 0) // if negative or 2 unused bits are not 0.
```

**Pull request merged**